devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/signal-graph/devtools-signal-graph.ts

```diff
@@ -0,0 +1,155 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import {DebugSignalGraph, DebugSignalGraphNode} from '../../../../../../protocol';
+import {
+  DevtoolsClusterNodeType,
+  DevtoolsSignalGraph,
+  DevtoolsSignalGraphNode,
+} from './signal-graph-types';
+import {checkClusterMatch} from './utils';
+
+interface Cluster {
+  id: string;
+  type: DevtoolsClusterNodeType;
+  nodes: Set<string>;
+  producers: Set<number>;
+  consumers: Set<number>;
+  name: string;
+  previewNode?: number;
+}
+
+type ClusterIdentifier = (nodes: DebugSignalGraph) => Cluster[];
+
+// Finds all `resource` clusters in the graph.
+const resourceClusterIdentifier: ClusterIdentifier = (graph) => {
+  const clusters: Map<string, Cluster> = new Map();
+
+  const isNodePartOfCluster = (n: DebugSignalGraphNode, name: string) => {
+    const match = checkClusterMatch(n);
+    return match && match.clusterName === name;
+  };
+
+  for (let i = 0; i < graph.nodes.length; i++) {
+    const node = graph.nodes[i];
+    const match = checkClusterMatch(node);
+    if (!match) {
+      continue;
+    }
+
+    const name = match.clusterName;
+    let cluster = clusters.get(name);
+    if (!cluster) {
+      cluster = {
+        id: `cl_${name}`,
+        type: 'resource',
+        name,
+        consumers: new Set(),
+        producers: new Set(),
+        nodes: new Set(),
+      };
+      clusters.set(name, cluster);
+    }
+
+    cluster.nodes.add(node.id);
+
+    for (const edge of graph.edges) {
+      // Note that we have to make sure that the consumer is not part of the
+      // same cluster since we'll end up with a circular dependency.
+      if (edge.producer === i && !isNodePartOfCluster(graph.nodes[edge.consumer], name)) {
+        cluster.consumers.add(edge.consumer);
+      }
+
+      // Same for producers
+      if (edge.consumer === i && !isNodePartOfCluster(graph.nodes[edge.producer], name)) {
+        cluster.producers.add(edge.producer);
+      }
+    }
+
+    if (match.signalName === 'value') {
+      cluster.previewNode = i;
+    }
+  }
+
+  return [...clusters].map(([, cluster]) => cluster);
```

Comment:
Nit: Prefer `Array.from`, you can also inline the map to avoid allocating an intermediary array.

https://techhub.social/@develwithoutacause/111819710786873448

```suggestion
  return Array.from(clusters, ([, cluster]) => cluster);
```

Alternatively,

```suggestion
  return Array.from(clusters.values());
```